dask/base.py

```diff
@@ -0,0 +1,85 @@
+import warnings
+from operator import attrgetter
+
+from toolz import merge, groupby, unique
+
+from .context import _globals
+
+
+class Base(object):
+    """Base class for dask collections"""
+
+    def visualize(self, optimize_graph=False):
+        from dask.dot import dot_graph
+        if optimize_graph:
+            return dot_graph(self._config.optimize(self.dask, self._keys()))
+        else:
+            return dot_graph(self.dask)
+
+    def _visualize(self, optimize_graph=False):
+        warn = DeprecationWarning("``_visualize`` is deprecated, use "
+                                  "``visualize`` instead.")
+        warnings.warn(warn)
+        return self.visualize(optimize_graph)
+
+    def compute(self, **kwargs):
+        return compute(self, **kwargs)[0]
+
+
+class Config(object):
+    """Specifies a configuration for a dask collection.
+
+    This holds functions that are used to specialize dask schedulers for a
+    specific collection.
+
+    Parameters
+    ----------
+    optimize : callable
+        Takes a dask and iterable of keys, and returns an optimized dask.
+    default_get : callable
+        The default scheduler to use (e.g. ``dask.threaded.get``)
+    finalize : callable
+        Takes a dask collection and an iterable of results, and converts them
+        into the desired python type.
+    """
+    def __init__(self, optimize, default_get, finalize):
+        self.optimize = optimize
+        self.default_get = default_get
+        self.finalize = finalize
```

Comment:
Sure. Sorry if I got a little heated there. Anyway, this whole PR solves is meant to solve two problems:

1. Unite the collections and remove repeated code. This now defines `get` and `compute` in one location instead of several duplicated versions.
2. Extract out the methods needed to integrate with `dask.imperative`. These are `finalize` and `optimize`. `default_get` was included just because it was the only other collection specific thing.

If you can solve all of those in a different way, please do. Personally, I think this approach simplifies things greatly, and I prefer grouping these functions together under one object that has docs describing what each function is for. I also like that I can subclass `Base`, add `_config` to the class, and not have to write my own `compute` method (and have it automatically play well with other classes/imperative). Another approach could be equally valid though.